core/arch/riscv64/riscv64.asm

```diff
@@ -110,7 +115,28 @@ call_dispatch_alt_stack_no_free:
  */
         DECLARE_EXPORTED_FUNC(dr_call_on_clean_stack)
 GLOBAL_LABEL(dr_call_on_clean_stack:)
-/* FIXME i#3544: Not implemented */
+        addi     sp, sp, -16
+        sd       ra, 8(sp)   /* Save link register. */
+        sd       s0, 0(sp)
+        mv       s0, sp      /* Save sp across the call. */
+        /* Swap stacks. */
+        ld       sp, dstack_OFFSET(ARG1)
+        /* Set up args. */
+        mv       ra,   ARG2 /* void *(*func)(arg1...arg8) */
+        mv       ARG1, ARG3          /* void *arg1 */
+        mv       ARG2, ARG4          /* void *arg2 */
+        mv       ARG3, ARG5          /* void *arg3 */
+        mv       ARG4, ARG6          /* void *arg4 */
+        mv       ARG5, ARG7          /* void *arg5 */
+        mv       ARG6, ARG8          /* void *arg6 */
+        ld       ARG7, 0(s0)         /* void *arg7, retrived from the stack */
```

Comment:
If I'm understand this correctly, arg7 and arg8 are expected to be passed on stack to dr_call_on_clean_stack. Here we're retrieving them from the stack top. But doesn't the stack top actually contain ra and s0 since we pushed them above?

Reply:
Yes, I think you're right, we should retrieve them from `16(s0)` and `24(s0)`.